src/ILCompiler.ReadyToRun/src/Compiler/ReadyToRunMetadataFieldLayoutAlgorithm.cs

```diff
@@ -4,44 +4,257 @@
+                };
+                Dictionary<DefType, FieldAndOffset[]> typeToFieldMap = new Dictionary<DefType, FieldAndOffset[]>();
+
+                for (int typeIndex = 1; typeIndex <= typeCountInModule; typeIndex++)
+                {
+                    List<FieldAndOffset> fieldsForType = null;
+                    DefType defType = (DefType)module.GetObject(MetadataTokens.TypeDefinitionHandle(typeIndex));
+                    if (defType.HasInstantiation)
+                    {
+                        // Generic types are exempt from the static field layout algorithm, see
+                        // <a href="https://github.com/dotnet/coreclr/blob/master/src/vm/ceeload.cpp#L2049">this check</a>.
+                        continue;
+                    }
+                    foreach (FieldDesc field in defType.GetFields())
+                    {
+                        if (field.IsStatic && !field.IsLiteral)
+                        {
+                            int index = (field.IsThreadStatic ? 1 : 0);
+                            int alignment = 1;
+                            int size = 0;
+                            bool isGcField = false;
+
+                            switch (field.FieldType.UnderlyingType.Category)
+                            {
+                                case TypeFlags.Byte:
+                                case TypeFlags.SByte:
+                                case TypeFlags.Boolean:
+                                    size = 1;
+                                    break;
+
+                                case TypeFlags.Int16:
+                                case TypeFlags.UInt16:
+                                case TypeFlags.Char:
+                                    alignment = 2;
+                                    size = 2;
+                                    break;
+
+                                case TypeFlags.Int32:
+                                case TypeFlags.UInt32:
+                                case TypeFlags.Single:
+                                    alignment = 4;
+                                    size = 4;
+                                    break;
+
+                                case TypeFlags.FunctionPointer:
+                                case TypeFlags.Pointer:
+                                case TypeFlags.IntPtr:
+                                case TypeFlags.UIntPtr:
+                                    alignment = _target.PointerSize;
+                                    size = _target.PointerSize;
+                                    break;
+
+                                case TypeFlags.Int64:
+                                case TypeFlags.UInt64:
+                                case TypeFlags.Double:
+                                    alignment = 8;
+                                    size = 8;
+                                    break;
+
+                                case TypeFlags.GenericParameter:
```

Comment:
ELEMENT_TYPE_VAR and ELEMENT_TYPE_MVAR correspond to SignatureTypeVariable and SignatureMethodVariable, if strictly comparing with CoreCLR source. Neither GenericParameter, nor the signature variables should show up here though because the type is not generic.